### PR TITLE
[toplevel] Refactor control loop

### PR DESCRIPTION
**Pull request by ejgallego**
We refactor control loop a bit to make the code more readable:

- the code for unhandled exception is not needed anymore as it cannot
  happen.
- we move the processing of toplevel commands to its own function
- we split away diff-specific functions.
- the PR now includes a fix for a bug that made the toplevel loop non tail recursive.

### CODE REVIEW

File: toplevel/coqloop.ml [hunk 1]
```diff
@@ -371,8 +371,89 @@ let exit_on_error =
   declare_bool_option_and_ref ~depr:false ~name:"coqtop-exit-on-error" ~key:["Coqtop";"Exit";"On";"Error"]
     ~value:false
 
+(* XXX: extension of Vernacentries.show_proof *)
+let show_proof_diff_to_pp pstate =
```
Comment (SkySkimmer) [inline toplevel/coqloop.ml]: Why XXX?

Comment (ejgallego) [reply toplevel/coqloop.ml]: This seems like a TODO from the PR that introduced this code, indeed seems pretty messy to duplicate here so this should be improved.

Once I have a bit of time I'll resume #8728 , or anyone else is welcome to do it; I believe there is room to cleanup that code.

Comment (gares) [reply toplevel/coqloop.ml]: It is not a TODO, it explains what the code does.
It may be an irrelevant comment, but how can it be a todo?

Comment (ejgallego) [reply toplevel/coqloop.ml]: I read it as "this is duplicate code, show_proof should be amended" but indeed I guess it depends on how interpret the original intention.

**Comment by ejgallego** [comment]
I've updated the comment, also rebased; needs assignee.

File: toplevel/coqloop.ml (hunk 1 above)
Comment (ejgallego) [reply toplevel/coqloop.ml]: Updated the comment, let me know how it looks know.

File: toplevel/coqloop.ml [hunk 2]
```diff
@@ -371,8 +371,90 @@ let exit_on_error =
...
+          n_pp
+        end
+    else
+      n_pp
+  with
+  | Pfedit.NoSuchGoal
+  | Option.IsNone ->
+    CErrors.user_err (str "No goals to show.")
+
+let process_stm ~state stm =
```
Comment (gares) [inline toplevel/coqloop.ml]: Please call this process_statement, since STM is already a thing

Comment (ejgallego) [reply toplevel/coqloop.ml]: fixed

**Comment by ejgallego** [comment]
@herbelin do you think you could process this? Thanks!

**Comment by herbelin** [comment]
> @herbelin do you think you could process this? Thanks!

Reviewing in details would require probably as much time as it took to be written, so I'm not very enthousiastic at doing this job. Maybe could we shortly discuss a bit in person and you explain me what is done?

**Comment by ejgallego** [comment]
> Maybe could we shortly discuss a bit in person and you explain me what is done?

Sure, thanks!

**Review by herbelin — APPROVED** [review]
This is a refactoring, with looks like a clear and balanced separation of the different steps, and the main loop changed so that it is now tail recursive. I like it.

Will merge tomorrow.

**Comment by ejgallego** [comment]
Merci @herbelin

**Pull request merged**